### PR TITLE
Map entities from a resource when written to the world.

### LINKED ISSUE #10958: `ReflectMapEntities` on Resources
## Bevy version

`0.12.1`

## What you did

```rust
#[derive(Resource, Reflect, Clone, Default)]
#[reflect(Resource, MapEntities)]
pub struct MyResource {
    entity: Entity
}

impl MapEntities for MyResource {
    fn map_entities(&mut self, entity_mapper: &mut EntityMapper) {
        *self.entity = entity_mapper.get_or_reserve(self.entity);
    }
}
```

## What went wrong

```
error[E0277]: the trait bound `app::MyResource: bevy::prelude::Component` is not satisfied
   --> src/main.rs:20:20
    |
500 | #[derive(Resource, Reflect, Clone, Default)]
    |                    ^^^^^^^ the trait `bevy::prelude::Component` is not implemented for `app::MyResource`
    |
    = help: the following other types implement trait `bevy::prelude::Component`:
              AccessibilityNode
              bevy::prelude::Name
              GlobalTransform
              Transform
              Children
              SceneInstance
              Parent
            and 170 others
    = note: required for `ReflectMapEntities` to implement `FromType<app::MyResource>`
    = note: this error originates in the derive macro `Reflect` (in Nightly builds, run with -Z macro-backtrace for more info)
```

## Additional Information

Looks like this is caused by `ReflectMapEntities` having a Component bound on `FromType`.

### PR DESCRIPTION
**Pull request by brandon-reinhart**
# Objective

- Fix #10958 by performing entity mapping on the entities inside of resources.

## Solution

- Resources can reflect(MapEntitiesResource) and impl MapEntities to get access to the mapper during the world insert of the scene.

## Testing

- A test resource_entity_map_maps_entities confirms the desired behavior.

## Changelog

- Added reflect(MapEntitiesResource) for mapping entities on Resources in a DynamicScene.

fixes 10958

### CODE REVIEW

File: crates/bevy_ecs/src/reflect/map_entities.rs [hunk 1]
```diff
@@ -73,3 +73,28 @@ impl<C: Component + MapEntities> FromType<C> for ReflectMapEntities {
         }
     }
 }
+
+#[derive(Clone)]
+pub struct ReflectMapEntitiesResource {
```
Comment (MrGVSV) [inline crates/bevy_ecs/src/reflect/map_entities.rs:78]: Do we need this type? I imagine `ReflectMapEntities` should be enough, right?

Comment (brandon-reinhart) [reply crates/bevy_ecs/src/reflect/map_entities.rs:78]: Gah! You scared the crap out of me. This comment popped up right in code while I was working.

Ah, no we probably don't. (But I wanted to separate this out while I was figuring out how this worked.)

Comment (brandon-reinhart) [reply crates/bevy_ecs/src/reflect/map_entities.rs:78]: The issue, at the moment, is the C: Component constraint on FromType<C>. I'm not sure how to work around that in the same type.

Comment (brandon-reinhart) [reply crates/bevy_ecs/src/reflect/map_entities.rs:78]: Ah I think I see.

Comment (MrGVSV) [reply crates/bevy_ecs/src/reflect/map_entities.rs:78]: > Gah! You scared the crap out of me. This comment popped up right in code while I was working.

Lol I didn't even know that was a thing 😂

> The issue, at the moment, is the C: Component constraint on FromType. I'm not sure how to work around that in the same type.

Oh that's unfortunate. Based on the impl, do you think we'd be able to merge them somehow?

If not, we might be forced to use a new type here (like you have already)

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/reflect/map_entities.rs:83]: I don't expect that merging to be possible: we need the bound in order to be able to extract the resource and modify it.

**Review by alice-i-cecile — APPROVED** [review]
Useful functionality, tightly scoped, with tests to validate that it works. Code quality is fine, you have my approval.

Like I said in the comment, I think we're out of luck for merging the two related types.

**Comment by alice-i-cecile** [comment]
I've added the `S-Needs-Release-Note` label here: I think that this is a high enough impact bug / limitation that it's worth calling out in a simple note so people can remove their workarounds.

**Review by MrGVSV — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1330 if you'd like to help out.

**Pull request merged**